### PR TITLE
front: router: add id support on EEX routes, need to replace modal cont…

### LINKED ISSUE #3661: front: new routing system (slug)
### Description and goal

We should have the project, study, scenario ids in the url:
`project/{project_id}/study/{study_id}/scenario/{scenario_id}`

Benefits
- avoid bugs with local storage
- allow users to send a direct link to a scenario
- separation stdcm/eex

### Acceptance criteria

- [ ] user is able to send a direct link to a scenario

### PR DESCRIPTION
**Pull request by Math-R**
…ext by react createPortal hook

close #3661

### CODE REVIEW

File: front/src/applications/operationalStudies/components/BreadCrumbs.tsx [hunk 1]
```diff
@@ -1,41 +1,44 @@
...
+  scenario?: ScenarioResult;
 };
 
-export default function BreadCrumbs({ projectName, studyName, scenarioName }: Props) {
+export default function BreadCrumbs({ project, study, scenario }: Props) {
   const { t } = useTranslation('operationalStudies/project');
   return (
     <div className="navbar-breadcrumbs">
-      {!projectName && !studyName && !scenarioName ? (
+      {!project && !study && !scenario ? (
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/BreadCrumbs.tsx:16]: The conditions are quite redundant and hard to read, do you want to refactor this here or should I open a new ticket to refactor it ?

Here is an example of rendering more explicit I think:
```
<div className="navbar-breadcrumbs">
      {!project ? (
        t('projectsList')
      ) : (
        <>
          <Link to="/operational-studies/projects">{t('projectsList')}</Link>
          <i className="icons-arrow-next icons-size-x75 text-muted" />
          {!study ? (
            project.name
          ) : (
            <>
              <Link to={`/operational-studies/projects/${project.id}`}>{project.name}</Link>
              <i className="icons-arrow-next icons-size-x75 text-muted" />
              {!scenario ? (
                study.name
              ) : (
                <>
                  <Link to={`/operational-studies/projects/${project.id}/studies/${study.id}`}>
                    {study.name}
                  </Link>
                  <i className="icons-arrow-next icons-size-x75 text-muted" />
                  {scenario.name}
                </>
              )}
            </>
          )}
        </>
      )}
    </div>
```

File: front/src/applications/operationalStudies/views/Project.tsx [hunk 2]
```diff
@@ -68,7 +67,7 @@ export default function Project() {
   const [imageUrl, setImageUrl] = useState('');
   const dispatch = useDispatch();
   const navigate = useNavigate();
-  const projectId = useSelector(getProjectID);
+  const { projectId } = useParams();
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/Project.tsx:65]: I think it's not a good thing to cast it to number and directly use it without any test. Maybe we can do this :
```suggestion
  const { projectId: urlProjectId } = useParams();
  const projectId = useMemo(() => Number(urlProjectId) || INVALID_ID, [urlProjectId]); // set INVALID_ID = -1 outside the component
```

File: front/src/applications/operationalStudies/views/Project.tsx [hunk 3]
```diff
@@ -84,7 +83,7 @@ export default function Project() {
   ];
 
   const { data: project, isError: isProjectError } = osrdEditoastApi.useGetProjectsByProjectIdQuery(
-    { projectId: projectId as number },
+    { projectId: Number(projectId) },
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/Project.tsx:81]: Then you need to check everywhere that projectId is not invalid
```suggestion
    { projectId },
    {
      skip: projectId === INVALID_ID,
    }
```

Also, you need to update the useEffect which redirect the user if the id is invalid. You can test this case by setting the id to 'a' for instance
```
  useEffect(() => {
    if (projectId === INVALID_ID) {
      navigate('/operational-studies/projects');
    }
  }, [projectId]);
```

File: front/package.json [hunk 4]
```diff
@@ -71,8 +71,10 @@
     "jwt-decode": "^3.1.2",
     "keycloak-js": "^20.0.3",
     "license-checker-rseidelsohn": "~3.3.0",
+    "localforage": "^1.10.0",
```
Comment (clarani) [inline front/package.json:74]: why did you install these librairies ? localforage, match-sorter and sort-by

**Review by clarani — CHANGES_REQUESTED** [review]
I left some comments, can you apply these changes in Project but also Study and Scenario ? Also you have 4 warnings and some functions are commented in ProjectCard, but it breaks a behaviour

File: front/package.json (hunk 4 above)
Comment (clarani) [reply front/package.json:74]: Specified in react-router-dom documentation

File: front/src/applications/operationalStudies/components/BreadCrumbs.tsx (hunk 1 above)
Comment (Yohh) [reply front/src/applications/operationalStudies/components/BreadCrumbs.tsx:16]: In my opinion nested ternary is quite more difficult to read, not sure this is the best idea

Comment (clarani) [reply front/src/applications/operationalStudies/components/BreadCrumbs.tsx:16]: Ok fine :)

File: front/src/applications/operationalStudies/views/Scenario.tsx [hunk 5]
```diff
@@ -129,6 +135,15 @@ export default function Scenario() {
       { skip: !timetableId }
     );
 
+  useEffect(() => {
+    if (!projectId || !studyId || !scenarioId) navigate('/operational-studies/projects');
+    // redirect if projectId or studyId is not a number
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/Scenario.tsx:140]: Move this comment above the useEffet and remove the lines 148/149
```suggestion
    // redirect if projectId, studyId or scenarioId is not a number
```

**Review by clarani — APPROVED** [review]
LGTM, tested :white_check_mark:

**Pull request merged**